core/include/vecmem/memory/unique_ptr.hpp

```diff
@@ -0,0 +1,462 @@
+ */
+template <typename T,
+          std::enable_if_t<!std::is_array_v<T> ||
+                               (std::is_array_v<T> && std::extent_v<T> == 0),
+                           bool> = true>
+struct unique_obj_deleter {
+public:
+    using pointer_t =
+        std::conditional_t<std::is_array_v<T>, std::decay_t<T>, T*>;
+    using storage_t = std::remove_pointer_t<pointer_t>;
+
+    /**
+     * @brief Default-construct a new unique object deleter.
+     *
+     * Having a default constructor for this class is not ideal, as having an
+     * empty deleter here doesn't really make any sense. However, if the
+     * deleter class is not default-constructible, then neither is the
+     * std::unique_ptr class using it. For that reason, we want it here.
+     */
+    unique_obj_deleter(void) = default;
+
+    /**
+     * @brief Construct a new unique object deleter.
+     *
+     * This is the preferred constructor for this type. It will store a memory
+     * resource, a pointer, a size, and possible an alignment and a number of
+     * elements.
+     *
+     * @param mr The memory resource to use for deallocation.
+     * @param p The pointer which we will want to deallocate.
+     * @param s The size of the allocation.
+     * @param a The alignment of the allocation, with 0 representing no
+     * alignment.
+     * @param n The number of elements in the allocation.
+     */
+    unique_obj_deleter(memory_resource& mr, pointer_t p, std::size_t s,
+                       std::size_t a = 0, std::size_t n = 1)
+        : m_mr(&mr), m_ptr(p), m_size(s), m_align(a), m_elems(n) {}
+
+    /**
+     * @brief Copy a unique object deleter.
+     *
+     * @param i The object deleter to copy.
+     */
+    unique_obj_deleter(const unique_obj_deleter<T>& i)
+        : m_mr(i.m_mr),
+          m_ptr(i.m_ptr),
+          m_size(i.m_size),
+          m_align(i.m_align),
+          m_elems(i.m_elems) {}
+
+    /**
+     * @brief Copy a unique object deleter.
+     *
+     * @param i The object deleter to copy.
+     */
+    unique_obj_deleter(unique_obj_deleter<T>& i)
+        : m_mr(i.m_mr),
+          m_ptr(i.m_ptr),
+          m_size(i.m_size),
+          m_align(i.m_align),
+          m_elems(i.m_elems) {}
+
+    /**
+     * @brief Move a unique object deleter.
+     *
+     * @param i The object deleter to move.
+     */
+    unique_obj_deleter(unique_obj_deleter<T>&& i)
+        : m_mr(i.m_mr),
+          m_ptr(i.m_ptr),
+          m_size(i.m_size),
+          m_align(i.m_align),
+          m_elems(i.m_elems) {}
+
+    /**
+     * @brief Activate the deletion mechanism of the deleter.
+     *
+     * @param p The pointer to deallocate.
+     *
+     * @note The pointer argument to this function is in some ways obsolete,
+     * because we store the pointer as a member variable. However, having both
+     * makes it easier to cross-check the logic.
+     */
+    void operator()(pointer_t p) const {
+        assert(p == m_ptr);
```

Comment:
The `m_ptr` variable is really only used for this one line. Is it worth to declare it just for this? :thinking:

Reply:
Sure, I'll get rid of it.